### PR TITLE
gcp spanner support added

### PR DESCRIPTION
**Pull request by GauravChawda**

### CODE REVIEW

File: tests/integration/cartography/intel/gcp/test_spanner.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+    nodes = neo4j_session.run(
+        """
+        MATCH (r:GCPSpannerInstanceBackup) return r.name
+        """,
+    )
+
+    actual_nodes = {n['r.name'] for n in nodes}
+
+    assert actual_nodes == expected_nodes
```
Comment (ajinkya1986) [inline tests/integration/cartography/intel/gcp/test_spanner.py:100]: Add tests to check relationships between different resources. Like `instance` and `database`. `backup` and `database`

**Review by ajinkya1986 — COMMENTED** [review]
Add tests for relationships between nodes.

File: cartography/data/indexes.cypher [hunk 2]
```diff
@@ -162,6 +162,10 @@ CREATE INDEX IF NOT EXISTS FOR (n:GCPFirestoreDatabase) ON (n.id);
 CREATE INDEX IF NOT EXISTS FOR (n:GCPFirestoreIndex) ON (n.id);
 CREATE INDEX IF NOT EXISTS FOR (n:GCPBigqueryDataset) ON (n.id);
 CREATE INDEX IF NOT EXISTS FOR (n:GCPBigqueryTable) ON (n.id);
+CREATE INDEX IF NOT EXISTS FOR (n:GCPSpannerInstance) ON (n.name);
+CREATE INDEX IF NOT EXISTS FOR (n:GCPSpannerInstanceConfig) ON (n.name);
+CREATE INDEX IF NOT EXISTS FOR (n:GCPSpannerInstanceBackup) ON (n.name);
+CREATE INDEX IF NOT EXISTS FOR (n:GCPSpannerInstanceDatabase) ON (n.name);
```
Comment (ganeshjagdale9545) [inline cartography/data/indexes.cypher:168]: create indexes on id

File: cartography/intel/gcp/spanner.py [hunk 3]
```diff
@@ -0,0 +1,417 @@
...
+@timeit
+def load_spanner_instances(session: neo4j.Session, data_list: List[Dict], project_id: str, update_tag: int) -> None:
+    session.write_transaction(load_spanner_instances_tx, data_list, project_id, update_tag)
+
+
+@timeit
+def load_spanner_instances_tx(tx: neo4j.Transaction, data: List[Dict], project_id: str, gcp_update_tag: int) -> None:
+    query = """
+    UNWIND $Records as record
+    MERGE (instance:GCPSpannerInstance{name:record.name})
```
Comment (ganeshjagdale9545) [inline cartography/intel/gcp/spanner.py:65]: load `name` as id

File: cartography/intel/gcp/spanner.py [hunk 4]
```diff
@@ -0,0 +1,417 @@
...
+    ON CREATE SET
+        instance.firstseen = timestamp()
+    SET
+        instance.lastupdated = $gcp_update_tag,
+        instance.region = record.region,
+        instance.name = record.name,
```
Comment (ganeshjagdale9545) [inline cartography/intel/gcp/spanner.py:71]: name field of api have format `projects/<project>/instances/[a-z][-a-z0-9]*[a-z0-9]`. we load name field as a name of instance only

File: cartography/intel/gcp/spanner.py [hunk 5]
```diff
@@ -0,0 +1,417 @@
...
+@timeit
+def load_spanner_instance_configs(session: neo4j.Session, data_list: List[Dict], project_id: str, update_tag: int) -> None:
+    session.write_transaction(load_spanner_instance_configs_tx, data_list, project_id, update_tag)
+
+
+@timeit
+def load_spanner_instance_configs_tx(tx: neo4j.Transaction, data: List[Dict], project_id: str, gcp_update_tag: int) -> None:
+    query = """
+    UNWIND $Records as record
+    MERGE (instance_config:GCPSpannerInstanceConfig{name: record.name})
```
Comment (ganeshjagdale9545) [inline cartography/intel/gcp/spanner.py:141]: load name as id

File: cartography/intel/gcp/spanner.py [hunk 6]
```diff
@@ -0,0 +1,417 @@
...
+    ON CREATE SET
+        instance_config.firstseen = timestamp()
+    SET
+        instance_config.lastupdated = $gcp_update_tag,
+        instance_config.region = record.region,
+        instance_config.name = record.name,
```
Comment (ganeshjagdale9545) [inline cartography/intel/gcp/spanner.py:147]: same as above

File: cartography/intel/gcp/spanner.py [hunk 7]
```diff
@@ -0,0 +1,417 @@
...
+@timeit
+def load_spanner_instance_configs_replicas(session: neo4j.Session, data_list: List[Dict], project_id: str, update_tag: int) -> None:
+    session.write_transaction(load_spanner_instance_configs_replicas_tx, data_list, project_id, update_tag)
+
+
+@timeit
+def load_spanner_instance_configs_replicas_tx(tx: neo4j.Transaction, data: List[Dict], project_id: str, gcp_update_tag: int) -> None:
+    query = """
+    UNWIND $Records as record
+    MERGE (replica:GCPSpannerInstanceConfigReplica{default_leader_location: record.defaultLeaderLocation, location: record.location, type: record.type})
```
Comment (ganeshjagdale9545) [inline cartography/intel/gcp/spanner.py:189]: create id for replica in the format `instance_config_id/replica_location`

Comment (ganeshjagdale9545) [reply cartography/intel/gcp/spanner.py:189]: here load only `id`

File: cartography/intel/gcp/spanner.py [hunk 8]
```diff
@@ -0,0 +1,417 @@
...
+@timeit
+def load_spanner_instances_databases(session: neo4j.Session, data_list: List[Dict], project_id: str, update_tag: int) -> None:
+    session.write_transaction(load_spanner_instances_databases_tx, data_list, project_id, update_tag)
+
+
+@timeit
+def load_spanner_instances_databases_tx(tx: neo4j.Transaction, data: List[Dict], project_id: str, gcp_update_tag: int) -> None:
+    query = """
+    UNWIND $Records as record
+    MERGE (database:GCPSpannerInstanceDatabase{name: record.name})
```
Comment (ganeshjagdale9545) [inline cartography/intel/gcp/spanner.py:258]: load name as id

File: cartography/intel/gcp/spanner.py [hunk 9]
```diff
@@ -0,0 +1,417 @@
...
+    ON CREATE SET
+        database.firstseen = timestamp()
+    SET
+        database.lastupdated = $gcp_update_tag,
+        database.region = record.region,
+        database.name = record.name,
```
Comment (ganeshjagdale9545) [inline cartography/intel/gcp/spanner.py:264]: name should be a display name take it from name field of api

File: cartography/intel/gcp/spanner.py [hunk 10]
```diff
@@ -0,0 +1,417 @@
...
+@timeit
+def load_spanner_instances_backups(session: neo4j.Session, data_list: List[Dict], project_id: str, update_tag: int) -> None:
+    session.write_transaction(load_spanner_instances_backups_tx, data_list, project_id, update_tag)
+
+
+@timeit
+def load_spanner_instances_backups_tx(tx: neo4j.Transaction, data: List[Dict], project_id: str, gcp_update_tag: int) -> None:
+    query = """
+    UNWIND $Records as record
+    MERGE (backup:GCPSpannerInstanceBackup{name: record.name})
```
Comment (ganeshjagdale9545) [inline cartography/intel/gcp/spanner.py:341]: load name as id

File: cartography/intel/gcp/spanner.py [hunk 11]
```diff
@@ -0,0 +1,417 @@
...
+    ON CREATE SET
+        backup.firstseen = timestamp()
+    SET
+        backup.lastupdated = $gcp_update_tag,
+        backup.region = record.region,
+        backup.name = record.name,
```
Comment (ganeshjagdale9545) [inline cartography/intel/gcp/spanner.py:347]: same as above

File: cartography/intel/gcp/spanner.py [hunk 12]
```diff
@@ -0,0 +1,417 @@
...
+        backup.expire_time = record.expireTime,
+        backup.max_expire_time = record.maxExpireTime,
+        backup.referencing_backups = record.referencingBackups,
+        backup.referencing_databases = record.referencingDatabases,
+        backup.size_bytes = record.sizeBytes,
+        backup.state = record.state,
+        backup.version_time = record.versionTime
+    WITH backup, record
+    MATCH (database:GCPSpannerInstanceDatabase{name: record.database})
+    MERGE (backup)-[r:OF_DATABASE]->(database)
```
Comment (ganeshjagdale9545) [inline cartography/intel/gcp/spanner.py:362]: also create relationship between instance

**Review by ganeshjagdale9545 — CHANGES_REQUESTED** [review]
few minor changes

File: cartography/intel/gcp/spanner.py [hunk 13]
```diff
@@ -0,0 +1,429 @@
...
+
+
+@timeit
+def transform_instances(instances: List[Dict], project_id: str) -> List[Dict]:
+    transformed_instances = []
+    for instance in instances:
+        instance['project_id'] = project_id
+        instance['region'] = 'global'
+        instance['id'] = instance['name']
+        instance['instance_id'] = instance['name'].split('/')[-1]
```
Comment (ganeshjagdale9545) [inline cartography/intel/gcp/spanner.py:51]: `instance_id` rename to `name`

File: cartography/intel/gcp/spanner.py [hunk 14]
```diff
@@ -0,0 +1,429 @@
...
+    query = """
+    UNWIND $Records as record
+    MERGE (instance:GCPSpannerInstance{id:record.id})
+    ON CREATE SET
+        instance.firstseen = timestamp()
+    SET
+        instance.lastupdated = $gcp_update_tag,
+        instance.region = record.region,
+        instance.id = record.id,
+        instance.name = record.instance_id,
```
Comment (ganeshjagdale9545) [inline cartography/intel/gcp/spanner.py:73]: `instance_id` rename to `name`

File: cartography/intel/gcp/spanner.py [hunk 15]
```diff
@@ -0,0 +1,429 @@
...
+
+
+@timeit
+def transform_spanner_instance_configs_replicas(instance_configs: List[Dict], project_id: str, regions: list, common_job_parameters):
+    all_replicas = []
+    for instance_config in instance_configs:
+        replicas = instance_config.get('replicas', [])
+        for replica in replicas:
+            replica['config'] = instance_config.get('name')
+            replica['id'] = instance_config.get('name') + '/replicas/' + replica.get('location')
```
Comment (ganeshjagdale9545) [inline cartography/intel/gcp/spanner.py:180]: Instead of this use f string

File: cartography/intel/gcp/spanner.py [hunk 16]
```diff
@@ -0,0 +1,429 @@
...
+@timeit
+def load_spanner_instance_configs_replicas_tx(tx: neo4j.Transaction, data: List[Dict], project_id: str, gcp_update_tag: int) -> None:
+    query = """
+    UNWIND $Records as record
+    MERGE (replica:GCPSpannerInstanceConfigReplica{id: record.id})
+    ON CREATE SET
+        replica.firstseen = timestamp()
+    SET
+        replica.lastupdated = $gcp_update_tag,
+        replica.id = record.id,
```
Comment (ganeshjagdale9545) [inline cartography/intel/gcp/spanner.py:199]: load field once

File: cartography/intel/gcp/spanner.py [hunk 17]
```diff
@@ -0,0 +1,429 @@
...
+def load_spanner_instance_configs_tx(tx: neo4j.Transaction, data: List[Dict], project_id: str, gcp_update_tag: int) -> None:
+    query = """
+    UNWIND $Records as record
+    MERGE (instance_config:GCPSpannerInstanceConfig{id: record.id})
+    ON CREATE SET
+        instance_config.firstseen = timestamp()
+    SET
+        instance_config.lastupdated = $gcp_update_tag,
+        instance_config.region = record.region,
+        instance_config.id = record.id,
```
Comment (ganeshjagdale9545) [inline cartography/intel/gcp/spanner.py:150]: load field once

File: cartography/intel/gcp/spanner.py [hunk 18]
```diff
@@ -0,0 +1,429 @@
...
+@timeit
+def load_spanner_instances_tx(tx: neo4j.Transaction, data: List[Dict], project_id: str, gcp_update_tag: int) -> None:
+    query = """
+    UNWIND $Records as record
+    MERGE (instance:GCPSpannerInstance{id:record.id})
+    ON CREATE SET
+        instance.firstseen = timestamp()
+    SET
+        instance.lastupdated = $gcp_update_tag,
+        instance.region = record.region,
```
Comment (ganeshjagdale9545) [inline cartography/intel/gcp/spanner.py:71]: load field once

File: cartography/intel/gcp/spanner.py [hunk 19]
```diff
@@ -0,0 +1,429 @@
...
+        instance.id = record.id,
```
Comment (ganeshjagdale9545) [inline cartography/intel/gcp/spanner.py:72]: load field once

File: cartography/intel/gcp/spanner.py [hunk 20]
```diff
@@ -0,0 +1,429 @@
...
+
+@timeit
+def load_spanner_instances_databases(session: neo4j.Session, data_list: List[Dict], project_id: str, update_tag: int) -> None:
+    session.write_transaction(load_spanner_instances_databases_tx, data_list, project_id, update_tag)
+
+
+@timeit
+def load_spanner_instances_databases_tx(tx: neo4j.Transaction, data: List[Dict], project_id: str, gcp_update_tag: int) -> None:
+    for da in data:
+        print(f"found -> {da['restoreInfo']['backupInfo']['backup']}")
```
Comment (ganeshjagdale9545) [inline cartography/intel/gcp/spanner.py:264]: Remove print statement

File: cartography/intel/gcp/spanner.py [hunk 21]
```diff
@@ -0,0 +1,429 @@
...
+    query = """
+    UNWIND $Records as record
+    MERGE (database:GCPSpannerInstanceDatabase{id: record.id})
+    ON CREATE SET
+        database.firstseen = timestamp()
+    SET
+        database.lastupdated = $gcp_update_tag,
+        database.region = record.region,
+        database.id = record.id,
```
Comment (ganeshjagdale9545) [inline cartography/intel/gcp/spanner.py:273]: same as above

File: cartography/intel/gcp/spanner.py [hunk 22]
```diff
@@ -0,0 +1,429 @@
...
+def load_spanner_instances_backups_tx(tx: neo4j.Transaction, data: List[Dict], project_id: str, gcp_update_tag: int) -> None:
+    query = """
+    UNWIND $Records as record
+    MERGE (backup:GCPSpannerInstanceBackup{id: record.id})
+    ON CREATE SET
+        backup.firstseen = timestamp()
+    SET
+        backup.lastupdated = $gcp_update_tag,
+        backup.region = record.region,
+        backup.id = record.id,
```
Comment (ganeshjagdale9545) [inline cartography/intel/gcp/spanner.py:358]: same as above

**Review by ganeshjagdale9545 — CHANGES_REQUESTED** [review]
code overall looks good. few minor changes needed

**Review by ganeshjagdale9545 — APPROVED** [review]
looks good

**Review by mpurusottamc — APPROVED** [review]
looks good.

**Comment by mpurusottamc** [comment]
@GauravChawda can you resolve the conflicts? I will merge your code after that.

File: cartography/intel/gcp/__init__.py [hunk 23]
```diff
@@ -27,13 +27,13 @@
...
 Resources = namedtuple(
     'Resources', 'compute gke cloudfunction crm_v1 crm_v2 dns storage serviceusage \
-        iam apigateway cloudkms cloudrun sql bigtable firestore pubsub dataproc cloudmonitoring cloud_logging cloudcdn loadbalancer apikey bigquery spanner',
+        iam apigateway cloudkms cloudrun sql bigtable firestore pubsub dataproc cloudmonitoring cloud_logging cloudcdn loadbalancer apikey bigquery pubsublite spanner',
 )
 
 # Mapping of service short names to their full names as in docs. See https://developers.google.com/apis-explorer,
 # and https://cloud.google.com/service-usage/docs/reference/rest/v1/services#ServiceConfig
 Services = namedtuple(
-    'Services', 'compute storage gke dns cloudfunction crm_v1 crm_v2 cloudkms cloudrun iam apigateway sql bigtable firestore apikey bigquery spanner',
```
Comment (mpurusottamc) [inline cartography/intel/gcp/__init__.py]: did we by mistake remove spanner? I see few more areas where spanner is missing below.

File: cartography/intel/gcp/__init__.py [hunk 24]
```diff
@@ -52,7 +52,7 @@
     firestore='firestore.googleapis.com',
     apikey='apikeys.googleapis.com',
     bigquery='bigquery.googleapis.com',
-    spanner='spanner.googleapis.com'
```
Comment (mpurusottamc) [inline cartography/intel/gcp/__init__.py]: like this.

File: cartography/intel/gcp/__init__.py [hunk 25]
```diff
@@ -341,7 +331,6 @@ def _initialize_resources(credentials: GoogleCredentials) -> Resource:
         loadbalancer=_get_compute_resource(credentials),
         apikey=_get_apikey_resource(credentials),
         bigquery=_get_bigquery_resource(credentials),
-        spanner=_get_spanner_resource(credentials)
```
Comment (mpurusottamc) [inline cartography/intel/gcp/__init__.py]: here also

File: cartography/intel/gcp/__init__.py [hunk 26]
```diff
@@ -301,16 +301,6 @@ def _get_bigquery_resource(credentials: GoogleCredentials) -> Resource:
     return googleapiclient.discovery.build('bigquery', 'v2', credentials=credentials, cache_discovery=False)
 
 
-def _get_spanner_resource(credentials: GoogleCredentials) -> Resource:
```
Comment (mpurusottamc) [inline cartography/intel/gcp/__init__.py]: and here

File: cartography/intel/gcp/__init__.py [hunk 27]
```diff
@@ -27,13 +27,13 @@
...
 Resources = namedtuple(
     'Resources', 'compute gke cloudfunction crm_v1 crm_v2 dns storage serviceusage \
-        iam apigateway cloudkms cloudrun sql bigtable firestore pubsub dataproc cloudmonitoring cloud_logging cloudcdn loadbalancer apikey bigquery pubsublite',
+        iam apigateway cloudkms cloudrun sql bigtable firestore pubsub dataproc cloudmonitoring cloud_logging cloudcdn loadbalancer apikey bigquery pubsublite spanner',
 )
 
 # Mapping of service short names to their full names as in docs. See https://developers.google.com/apis-explorer,
 # and https://cloud.google.com/service-usage/docs/reference/rest/v1/services#ServiceConfig
 Services = namedtuple(
-    'Services', 'compute storage gke dns cloudfunction crm_v1 crm_v2 cloudkms cloudrun iam apigateway sql bigtable firestore apikey bigquery pubsublite',
```
Comment (mpurusottamc) [inline cartography/intel/gcp/__init__.py:36]: check this properly. pubsublite is also missing here.

File: cartography/intel/gcp/__init__.py [hunk 28]
```diff
@@ -52,7 +52,7 @@
     firestore='firestore.googleapis.com',
     apikey='apikeys.googleapis.com',
     bigquery='bigquery.googleapis.com',
-    pubsublite='pubsublite.googleapis.com',
```
Comment (mpurusottamc) [inline cartography/intel/gcp/__init__.py:55]: spanner and pubsublite both got removed as part of merge.

**Review by mpurusottamc — CHANGES_REQUESTED** [review]
please check all the code again. seems we migth have gotten rid of some spanner and pubsublite related code during the merge.

**Review by mpurusottamc — APPROVED** [review]
looks good.

**Pull request merged**